internal/controller/gateway.go

```diff
@@ -31,46 +37,201 @@ import (
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: gw.Namespace,
+						Name:      gw.Name,
+					},
+				},
+			}
+		}
+		r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+			Info("gateway not owned by contour, bypassing reconciliation.")
+		return []reconcile.Request{}
+	})
+}
+
+// classForGateway returns an nil, error if the gatewayclass referenced by gw does not exist
+// or nil, nil if gw is not owned by Contour. Otherwise, the referenced gatewayclass is returned.
+func (r *gatewayReconciler) classForGateway(gw *gatewayapi_v1alpha1.Gateway) (*gatewayapi_v1alpha1.GatewayClass, error) {
+	gc := &gatewayapi_v1alpha1.GatewayClass{}
+	if err := r.client.Get(r.ctx, types.NamespacedName{Name: gw.Spec.GatewayClassName}, gc); err != nil {
+		return nil, fmt.Errorf("failed to get gatewayclass %s: %w", gw.Spec.GatewayClassName, err)
+	}
+	if !r.isController(gc) {
+		return nil, nil
+	}
+	return gc, nil
+}
+
+// isController returns true if the controller of the provided gc matches Contour's
+// GatewayClass controller string.
+func (r *gatewayReconciler) isController(gc *gatewayapi_v1alpha1.GatewayClass) bool {
+	return gc.Spec.Controller == r.classController
+}
+
 func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log.WithField("namespace", request.Namespace).WithField("name", request.Name).Info("reconciling gateway")
 
-	// Fetch the Gateway from the cache.
-	gateway := &gatewayapi_v1alpha1.Gateway{}
-	err := r.client.Get(ctx, request.NamespacedName, gateway)
-	if errors.IsNotFound(err) {
-		r.eventHandler.OnDelete(&gatewayapi_v1alpha1.Gateway{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      request.Name,
-				Namespace: request.Namespace,
-			},
-		})
+	// Fetch the Gateway.
+	gw := &gatewayapi_v1alpha1.Gateway{}
+	if err := r.client.Get(ctx, request.NamespacedName, gw); err != nil {
+		if errors.IsNotFound(err) {
+			r.log.WithField("name", request.Name).WithField("namespace", request.Namespace).Info("failed to find gateway")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object, so requeue the request.
+		return reconcile.Result{}, fmt.Errorf("failed to get gateway %s/%s: %w", request.Namespace, request.Name, err)
+	}
+
+	// Check if object is deleted.
+	if !gw.ObjectMeta.DeletionTimestamp.IsZero() {
+		r.eventHandler.OnDelete(gw)
+		if err := r.ensureGatewayDeleted(gw); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to ensure deletion of gateway %s/%s: %w",
+				gw.Namespace, gw.Name, err)
+		}
 		return reconcile.Result{}, nil
 	}
 
-	// Pass the new changed object off to the eventHandler.
-	r.eventHandler.OnAdd(gateway)
+	// The gateway is safe to process, so check if it's valid.
+	errs := validation.ValidateGateway(ctx, r.client, gw)
+	if errs != nil {
+		r.log.WithField("name", request.Name).WithField("namespace", request.Namespace).
```

Comment:
We should also be setting the `Scheduled` Condition to `status: false` here, with a `NotReconciled` Reason.

Reply:
The Gateway has been reconciled, but no infra is being provisioned since it's invalid. I will set `Scheduled=false` and use an `InvalidGateway` reason with the message indicating why the Gateway is invalid.

Reply:
How does it work if not provisioning infra? Need to keep that use-case in mind.

Reply:
If not provisioning infra, `Scheduled` is roughly equivalent to the `Valid` condition we use on HTTPProxy, it indicates that whether the object is syntactically valid. Should this be more than one Condition? Possibly, but I wasn't able to make that argument convincingly enough upstream.